### PR TITLE
raftstore: check pending regions before step

### PR DESCRIPTION
**Pull request by BusyJay**
ref: https://github.com/pingcap/tikv/issues/866

### CODE REVIEW

File: src/raftstore/store/store.rs [hunk 1]
```diff
@@ -456,26 +458,35 @@ impl<T: Transport, C: PdClient> Store<T, C> {
         }
     }
 
-    fn is_snapshot_overlapped(&self, msg: &RaftMessage) -> Result<bool> {
+    fn is_snapshot_overlapped(&mut self, msg: &RaftMessage) -> Result<bool> {
```
Comment (ngaut) [inline src/raftstore/store/store.rs]: We can just return bool instead of Result<bool>

Comment (BusyJay) [inline src/raftstore/store/store.rs]: Decoding snapshot may fail.

File: src/raftstore/store/store.rs [hunk 2]
```diff
@@ -456,26 +458,35 @@ impl<T: Transport, C: PdClient> Store<T, C> {
...
+            let exist_region = self.region_peers[&exist_region_id].region();
+            if enc_start_key(exist_region) < enc_end_key(&snap_region) {
+                warn!("region overlapped {:?}, {:?}", exist_region, snap_region);
+                return Ok(true);
+            }
+        }
+        for region in &self.pending_regions {
+            if enc_start_key(region) < enc_end_key(&snap_region) &&
+               enc_end_key(region) > enc_start_key(&snap_region) {
+                return Ok(true);
```
Comment (siddontang) [inline src/raftstore/store/store.rs]: log here

File: tests/raftstore/transport_simulate.rs [hunk 3]
```diff
@@ -361,3 +362,53 @@ impl FilterFactory for DropSnapshot {
...
+                _ => false,
+            };
+            if paused {
+                self.dropped.compare_and_swap(false, true, Ordering::Relaxed);
+                pending_msg.push(m);
+            } else {
+                to_send.push(m);
+            }
+        }
+        if pending_msg.len() > 1 {
```
Comment (siddontang) [inline tests/raftstore/transport_simulate.rs]: if we send one snapshot, there is two pending msgs, one is MsgSnapshot and the other is SnapshotStatus, so here need > 2?

Comment (BusyJay) [inline tests/raftstore/transport_simulate.rs]: Each node has a different filter, hence MsgSnapshot and SnapshotStatus will appear in different filters.

**Comment by siddontang** [comment]
Rest LGTM

**Comment by ngaut** [comment]
LGTM

**Comment by BusyJay** [comment]
@siddontang PTAL

**Comment by siddontang** [comment]
LGTM

**Pull request merged**